libs/parser/parser.c

```diff
@@ -0,0 +1,425 @@
+	strcpy(p->headerInfos[id].headerName, headerName);
+
+	if (hmap_insert(p->headersMap,
+			p->headerInfos[id].headerName,
+			&(p->headerInfos[id])) == -1) {
+		fprintf(stderr, "Parser: it is impossible to define two headers with the same names\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+/* Returns a type of line from file */
+enum parser_lineType parser_lineTypeGet(char *line)
+{
+	enum parser_lineType result = type_InvalidLine;
+
+	if (regexec(&(parser_common.fieldRegex), line, 0, NULL, 0) == 0) {
+		result = type_Field;
+	}
+	else if (regexec(&(parser_common.commentRegex), line, 0, NULL, 0) == 0) {
+		result = type_Comment;
+	}
+	else if (regexec(&(parser_common.headerRegex), line, 0, NULL, 0) == 0) {
+		result = type_Header;
+	}
+
+	return result;
+}
+
+
+/* Returns next header name. If succeeded returns 0. */
+static int parser_headerGet(char *line, char *result)
+{
+	regmatch_t regmatch[HEADER_SUBEXPRESSIONS];
+	int len;
+	int err = -1;
+
+	if (regexec(&(parser_common.headerRegex), line, HEADER_SUBEXPRESSIONS, regmatch, 0) == 0) {
+		len = regmatch[1].rm_eo - regmatch[1].rm_so;
+
+		if (len <= MAX_HEADER_LEN) {
+			strncpy(result, &line[regmatch[1].rm_so], len);
+			result[len] = '\0';
+			err = 0;
+		}
+	}
+
+	return err;
+}
+
+
+/* Fills `parser_result_t` structure. If succeeded returns 0. */
+static int parser_fieldGet(char *line, parser_field_t *result)
+{
+	regmatch_t regmatch[FIELD_SUBEXPRESSIONS];
+	size_t fieldLen, valueLen;
+	int err = -1;
+
+	if (regexec(&(parser_common.fieldRegex), line, FIELD_SUBEXPRESSIONS, regmatch, 0) == 0) {
+		fieldLen = regmatch[1].rm_eo - regmatch[1].rm_so;
+		valueLen = regmatch[2].rm_eo - regmatch[2].rm_so;
+
+		if (fieldLen <= MAX_FIELD_LEN && valueLen <= MAX_VALUE_LEN) {
+			strncpy(result->field, &line[regmatch[1].rm_so], fieldLen);
+			strncpy(result->value, &line[regmatch[2].rm_so], valueLen);
+
+			result->field[fieldLen] = '\0';
+			result->value[valueLen] = '\0';
+
+			err = 0;
+		}
+	}
+
+	return err;
+}
+
+
+int parser_execute(parser_t *p, const char *path)
+{
+	FILE *file;
+
+	char *line = NULL;
+	size_t lineLen = 0;
+
+	char header[MAX_HEADER_LEN + 1];
+	parser_headerInfo_t *headerInfo = NULL;
+
+	int res, fieldsCnt = 0, err = 0;
+
+	if (p == NULL || path == NULL) {
+		fprintf(stderr, "Parser: invalid arguments. Only non NULL arguments are correct.\n");
+		return -1;
+	}
+
+	file = fopen(path, "r");
+	if (file == NULL) {
+		fprintf(stderr, "Parser: there was an error during opening the file %s\n", path);
+		return -1;
+	}
+
+	while (getline(&line, &lineLen, file) != -1) {
+		switch (parser_lineTypeGet(line)) {
+			case type_Header:
+				/* Invoking converter for the previous header. If this is first header, then `headerInfo` is NULL */
+				if (headerInfo != NULL) {
+					err = headerInfo->converter(p->fieldsMap);
+					hmap_clear(p->fieldsMap);
+					if (err != 0) {
+						fprintf(stderr, "Parser: there was an error during converter invocation\n");
```

Comment:
Please use imperative/nominal sentences as error messages (if possible). They are always shorter and straight informative.
```suggestion
						fprintf(stderr, "Parser: error on converter invocation\n");
```